ckanfunctionaltests/api/conftest.py

```diff
@@ -148,29 +154,96 @@ def _strip_unstable_data(obj):
         return obj
 
 
+def set_ckan_vars(json_data):
+    ckan_vars = {}
+    with open(f"ckan-vars.conf") as ckan_vars_file:
+        for line in ckan_vars_file:
+            name, val = line.partition("=")[::2]
+            ckan_vars[name.strip()] = val.strip()
+
+    str_data = json.dumps(json_data)
+
+    for key in ckan_vars:
+        str_data = str_data.replace(f'<<{key}>>', ckan_vars[key])
+    
+    return json.loads(str_data)
+
+
+_key_value_keys = ['harvest', 'extras']
+
+
+def clean_unstable_elements(json_data, parent=None, ignore_element="", is_key_value=True, clean_element=[]):
+    if isinstance(json_data, str):
+        return
+    for key in json_data.keys():
+        if key in ignore_element:
+            continue
+
+        if key in clean_element:
+            json_data[key] = f"<<{key}>>"
+        elif is_key_value and key in _key_value_keys:
+            for item in json_data[key]:
+                item_key = item["key"]
+                item["value"] = f"<<{item_key}-value>>"
+        elif isinstance(json_data[key], list):
+            for item in json_data[key]:
+                clean_unstable_elements(item, key, ignore_element)
+        elif isinstance(json_data[key], dict):
+            clean_unstable_elements(json_data[key], key, ignore_element)
+        elif key in _unstable_keys and not str(json_data[key]).startswith("<<"):
+            if not (key == "id" and not parent):
+                json_data[key] = f"<<{key}>>"
+    return json_data
```

Comment:
So.. from what I can tell, this has more or less duplicated the mechanism I created (using fuzzy equality, see `comparisons.py`) to be able to handle unstable data, but in a very odd way (which will break as soon as it involves any json-reserved characters or characters that will need escaping in json literals)

Reply:
I found that using the other mechanism made it more difficult to run checks against the data, this approach made it easier to identify what the problems with the data were.

Reply:
But running these tests against stable data doesn't test _anything_ more than the existing stable-data tests. If you **just** want to run tests against stable-data, just run the stable-data tests, and the entire test suite could have been done in about a quarter the number of LoC. I guess I should have put another config variable in `config.json` to allow you to do that.

Reply:
Well, things can be done to make the failure output of `comparisons.py` comparisons more obvious, but now we've got _two_ differently janky mechanisms sitting on top of each other both trying to do the same job and with no explanation of what's going on or why, and this one even invents its own config file format.

Reply:
This is an example of the difficulty in working out what the diffs are in the data when using the `AnySupersetOf` function -

```
 AssertionError: assert {'codelist': '',\n 'contact-email': '',\n 'contact-name': 'Example User',\n 'contact-phone': '',\n 'creator_user_id': 'fcf98421-cdd6-4028-9037-d329747004e5',\n 'foi-email': '',\n 'foi-name': '',\n 'foi-phone': '',\n 'foi-web': '',\n 'groups': [],\n 'id': 'a18d2811-13b0-4838-8bfb-5793433317b9',\n 'isopen': True,\n 'licence-custom': '',\n 'license_id': 'uk-ogl',\n 'license_title': 'UK Open Government Licence (OGL)',\n 'license_url': 'http://reference.data.gov.uk/id/open-government-licence',\n 'metadata_created': '2020-06-30T14:27:11.017278',\n 'metadata_modified': '2020-06-30T14:31:36.145781',\n 'name': 'example-dataset-number-one',\n 'notes': 'This is an example CKAN dataset consisting of a number of active '\n          'resources.',\n 'num_resources': 4,\n 'num_tags': 1,\n 'organization': {'approval_status': 'approved',\n                  'created': '2020-06-30T14:26:11.856464',\n                  'description': '',\n                  'id': '97078739-f34a-446f-aec8-99ac41b7cb7a',\n                  'image_url': '',\n                  'is_organization': True,\n                  'name': 'example-publisher-1',\n                  'revision_id': '38ed1bd8-d5e4-4513-b404-1b49e2d4ce9a',\n                  'state': 'active',\n                  'title': 'Example Publisher #1',\n                  'type': 'organization'},\n 'owner_org': '97078739-f34a-446f-aec8-99ac41b7cb7a',\n 'private': False,\n 'relationships_as_object': [],\n 'relationships_as_subject': [],\n 'revision_id': '5e8723a6-b814-4481-9b08-487793c2f88d',\n 'schema-vocabulary': '',\n 'state': 'active',\n 'theme-primary': '',\n 'title': 'Example Dataset #1',\n 'type': 'dataset',\n 'url': 'http://static-mock-harvest-source:11088/mock-third-party/example-dataset-1/about',\n 'version': None} == AnySupersetOfMapping({'owner_org': '04bc0fa7-ddd3-4ded-912d-d4de99df2924', 'contact-phone': '', 'relationships_as_object': AnySupersetOfSeq(()), 'private': False, 'num_tags': 1, 'foi-email': '', 'id': 'a18d2811-13b0-4838-8bfb-5793433317b9', 'metadata_created': '2020-06-24T13:33:13.660799', 'licence-custom': '', 'metadata_modified': '2020-06-24T13:50:17.328061', 'title': 'Example Dataset #1', 'state': 'active', 'version': None, 'license_id': 'uk-ogl', 'foi-web': '', 'num_resources': 4, 'contact-email': '', 'foi-name': '', 'groups': AnySupersetOfSeq(()), 'creator_user_id': '4734f6c9-bd43-4d7f-b1d3-2cd32981f961', 'relationships_as_subject': AnySupersetOfSeq(()), 'codelist': '', 'contact-name': 'Example User', 'name': 'example-dataset-number-one', 'isopen': True, 'schema-vocabulary': '', 'url': 'http://static-mock-harvest-source:11088/mock-third-party/example-dataset-1/about', 'type': 'dataset', 'notes': 'This is an example CKAN dataset consisting of a number of active resources.', 'license_title': 'UK Open Government Licence (OGL)', 'license_url': 'http://reference.data.gov.uk/id/open-government-licence', 'organization': AnySupersetOfMapping({'description': '', 'created': '2020-06-24T13:32:31.880826', 'title': 'Example Publisher #1', 'name': 'example-publisher-1', 'is_organization': True, 'state': 'active', 'image_url': '', 'revision_id': 'aeae631f-82c3-4ac8-ae53-ecbfe666c39d', 'type': 'organization', 'id': '04bc0fa7-ddd3-4ded-912d-d4de99df2924', 'approval_status': 'approved'}), 'revision_id': '8f42fae0-9209-4cbe-b9ba-6dec9c8b658b', 'foi-phone': '', 'theme-primary': ''})
E                 +{'owner_org': '97078739-f34a-446f-aec8-99ac41b7cb7a', 'contact-phone': '', 'relationships_as_object': [], 'private': False, 'num_tags': 1, 'foi-email': '', 'id': 'a18d2811-13b0-4838-8bfb-5793433317b9', 'metadata_created': '2020-06-30T14:27:11.017278', 'licence-custom': '', 'metadata_modified': '2020-06-30T14:31:36.145781', 'title': 'Example Dataset #1', 'state': 'active', 'version': None, 'license_id': 'uk-ogl', 'foi-web': '', 'num_resources': 4, 'contact-email': '', 'foi-name': '', 'groups': [], 'creator_user_id': 'fcf98421-cdd6-4028-9037-d329747004e5', 'relationships_as_subject': [], 'codelist': '', 'contact-name': 'Example User', 'name': 'example-dataset-number-one', 'isopen': True, 'schema-vocabulary': '', 'url': 'http://static-mock-harvest-source:11088/mock-third-party/example-dataset-1/about', 'type': 'dataset', 'notes': 'This is an example CKAN dataset consisting of a number of active resources.', 'license_title': 'UK Open Government Licence (OGL)', 'license_url': 'http://reference.data.gov.uk/id/open-government-licence', 'organization': {'description': '', 'created': '2020-06-30T14:26:11.856464', 'title': 'Example Publisher #1', 'name': 'example-publisher-1', 'is_organization': True, 'state': 'active', 'image_url': '', 'revision_id': '38ed1bd8-d5e4-4513-b404-1b49e2d4ce9a', 'type': 'organization', 'id': '97078739-f34a-446f-aec8-99ac41b7cb7a', 'approval_status': 'approved'}, 'revision_id': '5e8723a6-b814-4481-9b08-487793c2f88d', 'foi-phone': '', 'theme-primary': ''}
E                 -AnySupersetOfMapping({'owner_org': '04bc0fa7-ddd3-4ded-912d-d4de99df2924', 'contact-phone': '', 'relationships_as_object': AnySupersetOfSeq(()), 'private': False, 'num_tags': 1, 'foi-email': '', 'id': 'a18d2811-13b0-4838-8bfb-5793433317b9', 'metadata_created': '2020-06-24T13:33:13.660799', 'licence-custom': '', 'metadata_modified': '2020-06-24T13:50:17.328061', 'title': 'Example Dataset #1', 'state': 'active', 'version': None, 'license_id': 'uk-ogl', 'foi-web': '', 'num_resources': 4, 'contact-email': '', 'foi-name': '', 'groups': AnySupersetOfSeq(()), 'creator_user_id': '4734f6c9-bd43-4d7f-b1d3-2cd32981f961', 'relationships_as_subject': AnySupersetOfSeq(()), 'codelist': '', 'contact-name': 'Example User', 'name': 'example-dataset-number-one', 'isopen': True, 'schema-vocabulary': '', 'url': 'http://static-mock-harvest-source:11088/mock-third-party/example-dataset-1/about', 'type': 'dataset', 'notes': 'This is an example CKAN dataset consisting of a number of active resources.', 'license_title': 'UK Open Government Licence (OGL)', 'license_url': 'http://reference.data.gov.uk/id/open-government-licence', 'organization': AnySupersetOfMapping({'description': '', 'created': '2020-06-24T13:32:31.880826', 'title': 'Example Publisher #1', 'name': 'example-publisher-1', 'is_organization': True, 'state': 'active', 'image_url': '', 'revision_id': 'aeae631f-82c3-4ac8-ae53-ecbfe666c39d', 'type': 'organization', 'id': '04bc0fa7-ddd3-4ded-912d-d4de99df2924', 'approval_status': 'approved'}), 'revision_id': '8f42fae0-9209-4cbe-b9ba-6dec9c8b658b', 'foi-phone': '', 'theme-primary': ''})
```

by comparison when doing a standard comparison with the same test the output is clearer - 

```
E                 Differing items:
E                 {'revision_id': '5e8723a6-b814-4481-9b08-487793c2f88d'} != {'revision_id': '8f42fae0-9209-4cbe-b9ba-6dec9c8b658b'}
E                 {'creator_user_id': 'fcf98421-cdd6-4028-9037-d329747004e5'} != {'creator_user_id': '4734f6c9-bd43-4d7f-b1d3-2cd32981f961'}
E                 {'organization': {'approval_status': 'approved', 'created': '2020-06-30T14:26:11.856464', 'description': '', 'id': '97078739-f34a-446f-aec8-99ac41b7cb7a', ...}} != {'organization': {'approval_status': 'approved', 'created': '2020-06-24T13:32:31.880826', 'description': '', 'id': '04bc0fa7-ddd3-4ded-912d-d4de99df2924', ...}}
E                 {'owner_org': '97078739-f34a-446f-aec8-99ac41b7cb7a'} != {'owner_org': '04bc0fa7-ddd3-4ded-912d-d4de99df2924'}
E                 {'metadata_created': '2020-06-30T14:27:11.017278'} != {'metadata_created': '2020-06-24T13:33:13.660799'}
E                 {'metadata_modified': '2020-06-30T14:31:36.145781'} != {'metadata_modified': '2020-06-24T13:50:17.328061'}
E                 Full diff:
E                   {
E                    'codelist': '',
E                    'contact-email': '',
E                    'contact-name': 'Example User',
E                    'contact-phone': '',
E                 -  'creator_user_id': '4734f6c9-bd43-4d7f-b1d3-2cd32981f961',
E                 +  'creator_user_id': 'fcf98421-cdd6-4028-9037-d329747004e5',
E                    'foi-email': '',
E                    'foi-name': '',
E                    'foi-phone': '',
E                    'foi-web': '',
E                    'groups': [],
E                    'id': 'a18d2811-13b0-4838-8bfb-5793433317b9',
E                    'isopen': True,
E                    'licence-custom': '',
E                    'license_id': 'uk-ogl',
E                    'license_title': 'UK Open Government Licence (OGL)',
E                    'license_url': 'http://reference.data.gov.uk/id/open-government-licence',
E                 -  'metadata_created': '2020-06-24T13:33:13.660799',
E                 ?                               ^^  ^ ^^  ^ --  ^^
E                 +  'metadata_created': '2020-06-30T14:27:11.017278',
E                 ?                               ^^  ^ ^^  ^  + ^^^
E                 -  'metadata_modified': '2020-06-24T13:50:17.328061',
E                 ?                                ^^     --- --- --
E                 +  'metadata_modified': '2020-06-30T14:31:36.145781',
E                 ?                                ^^  ++ + +++++
E                    'name': 'example-dataset-number-one',
E                    'notes': 'This is an example CKAN dataset consisting of a number of active '
E                             'resources.',
E                    'num_resources': 4,
E                    'num_tags': 1,
E                    'organization': {'approval_status': 'approved',
E                 -                   'created': '2020-06-24T13:32:31.880826',
E                 ?                                       ^^  ^ -  ^   ^^^^
E                 +                   'created': '2020-06-30T14:26:11.856464',
E                 ?                                       ^^  ^  + ^   ^ +++
E                                     'description': '',
E                 -                   'id': '04bc0fa7-ddd3-4ded-912d-d4de99df2924',
E                 +                   'id': '97078739-f34a-446f-aec8-99ac41b7cb7a',
E                                     'image_url': '',
E                                     'is_organization': True,
E                                     'name': 'example-publisher-1',
E                 -                   'revision_id': 'aeae631f-82c3-4ac8-ae53-ecbfe666c39d',
E                 +                   'revision_id': '38ed1bd8-d5e4-4513-b404-1b49e2d4ce9a',
E                                     'state': 'active',
E                                     'title': 'Example Publisher #1',
E                                     'type': 'organization'},
E                 -  'owner_org': '04bc0fa7-ddd3-4ded-912d-d4de99df2924',
E                 +  'owner_org': '97078739-f34a-446f-aec8-99ac41b7cb7a',
E                    'private': False,
E                    'relationships_as_object': [],
E                    'relationships_as_subject': [],
E                 -  'revision_id': '8f42fae0-9209-4cbe-b9ba-6dec9c8b658b',
E                 +  'revision_id': '5e8723a6-b814-4481-9b08-487793c2f88d',
E                    'schema-vocabulary': '',
E                    'state': 'active',
E                    'theme-primary': '',
E                    'title': 'Example Dataset #1',
E                    'type': 'dataset',
E                    'url': 'http://static-mock-harvest-source:11088/mock-third-party/example-dataset-1/about',
E                    'version': None,
E                   }
```

I could spend some time figuring out how to improve the output, but my approach is less complicated and produces clearer output.

Reply:
All you have to do to make the diff output nicer is temporarily remove the `AnySupersetOf` wrapping, and you'll get an output very much like the above.

I also don't really agree that your approach is less complicated, especially seeing as we now have two overlapping methods of trying to achieve the same thing with no clear explanation why. If it's going to _replace_ the mechanisms I put in place, it should really replace them. It also introduces _another_ config file that needs to be filled out, with home brewed reading & parsing routines.

Reply:
Thats true, but also a bit odd to have to do something like that as part of the debugging process to identify any issues when there is a test failure between versions. If this becomes part of the normal testing process then it should be documented in the README.

The approach I've taken is less complicated in that the code I am using to prepare the data for comparison is a simpler approach than the `AnySuperSetOf` code which I found takes a bit of figuring out. 

I agree that having both running side by side is not ideal so I think there will probably be some further work to revisit the tests and take one approach or the other. 

As the aim of the card is to stabilise the tests for testing against the local dev stacks between CKAN version I think it's good enough for now.

Reply:
To you, your method might seem simpler, but to me I see a recursive function that does some string replacement stuff that could go wrong in numerous ways, has no comments or docstrings to explain its behaviour and has its own config file.

Ultimately though, you're going to be the one using this and you'll do what you like with it once I've left, so :shrug: .

I'd consider merging this to a separate branch though because I think the changes to the random -> stable data also make it less useful for testing production & staging. Case in point, I would _never_ have discovered the bugs I did in staging & production had I not been doing the randomized testing.

Reply:
Any bugs in the ckan or the functional test that were discovered through a randomised test should probably be added as additional tests if they haven't already, so that those areas are also covered. 

Rather than a separate branch, I'd rather just merge it in, as discussed I will create a card to refactor the test code to make it easier to maintain for some future work.